### ISSUE TITLE
Bayesian Hyperparameter Optimization

### ISSUE THREAD

**Issue by Riazone** [pycaret/pycaret#61]
Hi,
I was wondering if we can have Bayesian Hyperparameter Optimization technique used instead of Random Grid. This will help with speed of tuning and allow us to scrape through much larger grid scientifically.  We can have this enhancement along with ability to add custom grid in tuning.

Thanks

**Comment by pycaret** [comment]
Thanks.

Marking this as an enhancement.

**Comment by MattiaVerticchio** [comment]
It should be doable by using `skopt.BayesSearchCV` in place of `sklearn.model_selection.RandomizedSearchCV`.

Extract from [`scikit-optimized` homepage](https://scikit-optimize.github.io/stable/auto_examples/sklearn-gridsearchcv-replacement.html):

> Search for parameters of machine learning models that result in best cross-validation performance is necessary in almost all practical cases to get a model with best generalization estimate. A standard approach in scikit-learn is using `sklearn.model_selection.GridSearchCV` class, which takes a set of values for every parameter to try, and simply enumerates all combinations of parameter values. The complexity of such search grows exponentially with the addition of new parameters. A more scalable approach is using `sklearn.model_selection.RandomizedSearchCV`, which however does not take advantage of the structure of a search space.
> 
> Scikit-optimize provides a drop-in replacement for `sklearn.model_selection.GridSearchCV`, which utilizes Bayesian Optimization where a predictive model referred to as “surrogate” is used to model the search space and utilized to arrive at good parameter values combination as soon as possible.

**Comment by Yard1** [comment]
I tried to add it in with https://github.com/pycaret/pycaret/pull/306, but it's broken on the current scikit-learn version, and it doesn't seem like the project is very much active, so an update may not come any time soon.

**Comment by MattiaVerticchio** [comment]
What about [Optuna](https://optuna.org/)? It leverages Tree-based Parzen Estimators, which support categorical hyperparameters too.

**Comment by Yard1** [comment]
We're considering using Ray Tune, which IIRC also wraps Optuna.

https://github.com/pycaret/pycaret/issues/428

**Comment by Yard1** [comment]
Made a PR to `tune-sklearn` - https://github.com/ray-project/tune-sklearn/pull/68 - hopefully this gets accepted so we can use it without having to worry about maintaining our own fork.

**Comment by richardliaw** [comment]
@Yard1 just merged your patch - what are the other steps we need to take to make this work? :)

**Comment by Yard1** [comment]
@richardliaw Aw, thanks for your concern! We'll be working on implementing this in `2.2`, the next release after this. I'll make some proofs of concept before that, see how it all works, and if there are any problems, I'll make sure to let you know.

**Comment by Yard1** [comment]
@richardliaw I have tested patched tune-sklearn, and sadly it's not working as expected. It takes much longer than scikit-learn's RandomizedSearchCV (up to 5 times as much) and it runs out of memory easily on my laptop. It uses around 4-5 GBs of memory, compared to less than 100 MBs used by RandomizedSearchCV. This is all with default settings, running a slightly modified example. Is this expected behavior on personal computers? What can be done to improve this?

<details>
<summary>Code used</summary>

```python
"""
An example training a RandomForestClassifier, performing
randomized search using TuneSearchCV.
"""

from tune_sklearn import TuneSearchCV
from sklearn.ensemble import RandomForestClassifier
from sklearn import datasets
from sklearn.model_selection import train_test_split
from scipy.stats import randint
import numpy as np
from sklearn.model_selection import RandomizedSearchCV
import time

digits = datasets.load_digits()
x = digits.data
y = digits.target
x_train, x_test, y_train, y_test = train_test_split(x, y, test_size=.2)

clf = RandomForestClassifier()
param_distributions = {'n_estimators': [10, 20, 30, 40, 50, 60, 70, 80, 90, 100], 'criterion': ['gini', 'entropy'], 'max_depth': [10, 20, 30, 40, 50, 60, 70, 80, 90, 100, 110], 'min_samples_split': [2, 5, 7, 9, 10], 'min_samples_leaf': [1, 2, 4], 'max_features': ['auto', 'sqrt', 'log2'], 'bootstrap': [True, False]}

tune_search = TuneSearchCV(clf, param_distributions, n_iter=15)
start = time.time()
tune_search.fit(x_train, y_train)
end = time.time()
print("Tune Fit Time:", end - start)
pred = tune_search.predict(x_test)
accuracy = np.count_nonzero(np.array(pred) == np.array(y_test)) / len(pred)
print(accuracy)

tune_search = RandomizedSearchCV(clf, param_distributions, n_iter=15)
start = time.time()
tune_search.fit(x_train, y_train)
end = time.time()
print("Tune Fit Time:", end - start)
pred = tune_search.predict(x_test)
accuracy = np.count_nonzero(np.array(pred) == np.array(y_test)) / len(pred)
print(accuracy)
```
</details>

TuneSearchCV takes about 25 seconds on my machine, while RandomizedSearchCV takes around 7.

**Comment by richardliaw** [comment]
OK let me take a look into this!

**Comment by richardliaw** [comment]
Got it. @Yard1 it looks the parallelism is hurting us here. I switched Ray to use "local mode" (where we run everything in a single process), and I can get it to perform faster and better on my MacBook Pro:

Changes:

```diff
clf = RandomForestClassifier()
param_distributions = {'n_estimators': [10, 20, 30, 40, 50, 60, 70, 80, 90, 100], 'criterion': ['gini', 'entropy'], 'max_depth': [10, 20, 30, 40, 50, 60, 70, 80, 90, 100, 110], 'min_samples_split': [2, 5, 7, 9, 10], 'min_samples_leaf': [1, 2, 4], 'max_features': ['auto', 'sqrt', 'log2'], 'bootstrap': [True, False]}

+import ray
+ray.init(local_mode=True)
tune_search = TuneSearchCV(clf, param_distributions, n_iter=15)
```

Result:
```
(base) ➜  dev python test-sk.py
2020-08-20 10:54:21,618	INFO resource_spec.py:250 -- Starting Ray with 27.0 GiB memory available for workers and up to 13.52 GiB for objects. You can adjust these settings with ray.init(memory=<bytes>, object_store_memory=<bytes>).
2020-08-20 10:54:22,094	INFO services.py:1209 -- View the Ray dashboard at localhost:8265
/Users/rliaw/dev/tune-sklearn/tune_sklearn/tune_basesearch.py:249: UserWarning: Early stopping is not enabled. To enable early stopping, pass in a supported scheduler from Tune and ensure the estimator has `partial_fit`.
  warnings.warn("Early stopping is not enabled. "
2020-08-20 10:54:23,339	WARNING trainable.py:864 -- Trainable._setup is deprecated and will be removed in a future version of Ray. Override Trainable.setup instead.
2020-08-20 10:54:24,066	WARNING trainable.py:732 -- Trainable._train is deprecated and will be removed in a future version of Ray. Override Trainable.step instead.
2020-08-20 10:54:28,168	WARNING trainable.py:783 -- Trainable._save is deprecated and will be removed in a future version of Ray. Override Trainable.save_checkpoint instead.
Tune Fit Time: 5.584817886352539
0.9805555555555555
SK Fit Time: 9.139087915420532
0.975
```

Can you check if the memory still explodes on your side with this change? I could dig a bit into it if so.

We can also upstream this to tune-sklearn (as a flag). You would want `local_mode=True` for small models that train fast, but `local_mode=False` for massive parallelism of models that take >10 seconds to train each.

**Comment by Yard1** [comment]
@richardliaw Thank you very much. With the `local_mode` parameter set to True, memory usage is still bigger, but not nearly as much - 1 GB, which is manageable. It is also now faster than RandomizedSearchCV. It'd be great if that flag could be up streamed, it would help us here. @pycaret With this change, I think we will be able to include tune-sklearn as an option alongside Optuna, so the users will be able to choose which underlying library they prefer - or just stick to default RandomizedSearchCV.

**Comment by pycaret** [comment]
@richardliaw @Yard1 Thanks a lot for working on this.

Is the algorithm behind `optuna` and `tune-sklearn` same? I know `optuna` is bayesian search. What about `tune-sklearn`?

**Comment by richardliaw** [comment]
roughly speaking, it should be the same :) I think they implement TPE, which we provide via "hyperopt". We will also offer optuna underneath the hood.

**Comment by Yard1** [comment]
@pycaret tune-sklearn is a wrapper for several algorithms - originally it was only scikit-optimize, but with my patch it's also HpBandSter (BOHB), Hyperopt (Bayesian) and Optuna (yes, it's also wrapped). It also supports various early stopping schedulers. It's more configurable than Optuna due to more choices, but because it's running on Ray, it's more geared towards large datasets and distributed computing, while Optuna itself is more lightweight (and needs less stuff installed). They both have scikit-learn-like Optimizers, so we'd be able to use the same param grids and arguments with both, so it would be no issue to add support for both.

**Comment by pycaret** [comment]
@richardliaw Thanks a lot. Please apologize me for my ignorance. I have checked `ray` documentation and it looks really useful for our use-case here.

@Yard1 This makes good sense to me. We can extend as much as we want as long as we keep them in optional dependency with default behavior set to `random_search`. I like where this is going.

By the time we will be ready to drop this in `2.2` or `2.3` (depending on how refactoring goes), I think upstream change for `local_mode` in `tune-sklearn` would have been done by then. 

@richardliaw Thank you for your help. 
@Yard1 Great work as usual.

**Comment by koh-joshua** [comment]
+1 for Optuna. Really easy to set up and use. Works on Windows. The Optuna team is also very responsive and a very bunch of guys in general to work with. 
Not sure about Ray tune for Windows. I think it's only Linux for now?

**Comment by richardliaw** [comment]
Hm, Ray Tune should work fine for Windows and Mac. Let me know if you see otherwise!

**Comment by kmedved** [comment]
Optuna is quite nice and intuitive to use, but as far as I know, it does not support multiprocess local hyperparameter searching yet (I understand they are working on it). This is pretty important, especially for GBDT models, which frequently don't scale well with additional cores, and you're better off spending additional processing resources on more trials.

To the extent that's important for PyCaret, I would suggest Ray-Tune is a great solution.

**Comment by Yard1** [comment]
We will provide a choice of several libraries to be used in `tune_model()`.

**Comment by koh-joshua** [comment]
Really excited for Ray-Tune and Optuna to be part of tune_model() in the near future. Can't wait! Thank you, guys for your stellar effort in making PyCaret the go-to library for ML.

**Comment by pycaret** [comment]
@Riazone @MattiaVerticchio 

This issue has been fixed in the `pycaret-nightly` release. Please install pycaret-nightly using `pip install pycaret-nightly`. Feel free to open a new issue, in case the problem persists. This fix will be published to PyPi as `pycaret==2.2.0` on October 28, 2020.
